source/common/network/buffered_io_socket_handle_impl.cc

```diff
@@ -0,0 +1,272 @@
+                                 IoSocketError::deleteIoError)};
+    }
+  }
+  absl::FixedArray<iovec> iov(num_slice);
+  uint64_t num_slices_to_read = 0;
+  uint64_t bytes_to_read = 0;
+  for (; num_slices_to_read < num_slice && bytes_to_read < max_length; num_slices_to_read++) {
+    auto max_bytes_to_read = std::min(std::min(owned_buffer_.length(), max_length) - bytes_to_read,
+                                      uint64_t(slices[num_slices_to_read].len_));
+    owned_buffer_.copyOut(bytes_to_read, max_bytes_to_read, slices[num_slices_to_read].mem_);
+    bytes_to_read += max_bytes_to_read;
+  }
+  ASSERT(bytes_to_read <= max_length);
+  owned_buffer_.drain(bytes_to_read);
+  ENVOY_LOG(trace, "socket {} readv {} bytes", static_cast<void*>(this), bytes_to_read);
+  return {bytes_to_read, Api::IoErrorPtr(nullptr, [](Api::IoError*) {})};
+}
+
+Api::IoCallUint64Result BufferedIoSocketHandleImpl::read(Buffer::Instance& buffer,
+                                                         uint64_t max_length) {
+  if (!isOpen()) {
+    return {0,
+            Api::IoErrorPtr(new IoSocketError(SOCKET_ERROR_INVAL), IoSocketError::deleteIoError)};
+  }
+  if (owned_buffer_.length() == 0) {
+    if (read_end_stream_) {
+      return {0, Api::IoErrorPtr(nullptr, [](Api::IoError*) {})};
+    } else {
+      return {0, Api::IoErrorPtr(IoSocketError::getIoSocketEagainInstance(),
+                                 IoSocketError::deleteIoError)};
+    }
+  }
+  // TODO(lambdai): Move at slice boundary to move to reduce the copy.
+  uint64_t max_bytes_to_read = std::min(max_length, owned_buffer_.length());
+  buffer.move(owned_buffer_, max_bytes_to_read);
+  return {max_bytes_to_read, Api::IoErrorPtr(nullptr, [](Api::IoError*) {})};
+}
+
+Api::IoCallUint64Result BufferedIoSocketHandleImpl::writev(const Buffer::RawSlice* slices,
+                                                           uint64_t num_slice) {
+  if (!isOpen()) {
+    return {0,
+            Api::IoErrorPtr(new IoSocketError(SOCKET_ERROR_INVAL), IoSocketError::deleteIoError)};
+  }
+  if (!writable_peer_) {
+    return {0,
+            Api::IoErrorPtr(new IoSocketError(SOCKET_ERROR_INVAL), IoSocketError::deleteIoError)};
+  }
+  if (writable_peer_->isWriteEndSet()) {
+    // EPIPE or ENOTCONN
+    return {0,
+            Api::IoErrorPtr(new IoSocketError(SOCKET_ERROR_INVAL), IoSocketError::deleteIoError)};
+  }
+  if (!writable_peer_->isWritable()) {
+    return {0, Api::IoErrorPtr(IoSocketError::getIoSocketEagainInstance(),
+                               IoSocketError::deleteIoError)};
+  }
+  // Write along with iteration. Buffer guarantee the fragment is always append-able.
+  uint64_t total_bytes_to_write = 0;
+  for (uint64_t i = 0; i < num_slice; i++) {
+    if (slices[i].mem_ != nullptr && slices[i].len_ != 0) {
+      writable_peer_->getWriteBuffer()->add(slices[i].mem_, slices[i].len_);
+      total_bytes_to_write += slices[i].len_;
+    }
+  }
+  writable_peer_->maybeSetNewData();
```

Comment:
Perhaps I'm not reading the code correctly. I would assume that this should result in the `FileReadyCb` of the peer to be called with the `FileReadyType::Read` iff the owner of the `FileEvent` subscribed to the `FileReadyType::Rrite` events AND the peers buffer was empty, right? Otherwise the callback is not called.
In your implementation the callback is always called with the mask of watched events (without the Close). It does not sound right.

Reply:
>  iff the owner of the FileEvent subscribed to the FileReadyType::Rrite events AND the peers buffer was empty, right? 

The typo "FileReadyType::Rrite" I assume it is "Read" :)
The answer is No. Peer should be activated if the peer buffer is from "1 to 2" and "0 to 1". The maybe in "maybeSetNewData" referring to maybe the peer is interested in "read".

Is it a bad name? The name should be "always_set _new_data_but_maybe_raise_read_callback"

Reply:
Sorry, I'm not quite following you. What is "1 to 2" and "0 to 1"?
Also the function is not setting any data, it just triggers new event, right?

In any case I saw from the discussion with Antonio that it was intentional to pass all **watched** events to the event callback instead of just the event that was triggered. Was it for simplification? I suppose it is just the efficiency issue. However it should at least be commented. If fixing it is not too much code perhaps it should just be added into this PR.

Reply:
By the definition of level trigger, `pending data` from `no pending data to has pending data` should deliver an event to poller. Also `pending X bytes of data` to `pending X+Y bytes of data` should also deliver an event to poller. At least that is what epoll does :)

Deliver write even see read is just for simplicity. I can do it in the next round. 
I add the comment at DefaultEventListener::triggeredEvent()
```
// Return both read and write if enabled. Note that this implementation is inefficient. Read and
  // write events are supposed to be independent.
```
The EventListener need to watch more changes: buffer consuming, readDisable() and io handle state to filter out the unnecessary events.

Reply:
It was my understanding that `pending X bytes of data` to `pending X+Y bytes of data` will not deliver a new event in edge triggering mode. epoll documentation (which is also my experience) says that new events in edge triggering mode are delivered only after `read` returns EGAIN, i.e. you have to empty all pending data before new events are delivered.

Reply:
@yanavlasov I had this misunderstanding in the past. Until I work on this implementation :)
Regardless "X" to "X+Y" delivers an event in edge mode, since we never know Y bytes will arrive later, the only practical method is to `read until EAGAIN`.
Below is the simple epoll test case. I intentionally read byte by byte, and stop reading if the char is 'a'. 
https://gist.github.com/lambdai/a7f53965bebc771113ce2fc6376eb2c3#file-gistfile1-txt-L135-L139

1. I run nc to send "abc\n", the above program consume single char 'a' and  leave "bc\n" in buffer. (X= 3) 
2. At the nc, I send more characters "12345\n"".  (X =3, Y= 6). The program blocked at epoll_wait returns and EPOLLIN is triggered.

The output is as simple as 
```
$ ./a.out -s
[+] connected with 127.0.0.1:36896
[+] EPOLLIN event
[+] data: a
[+] EPOLLIN event
[+] data: b
[+] data: c
[+] data: 

[+] data: 1
[+] data: 2
[+] data: 3
[+] data: 4
[+] data: 5
[+] data: 
```